grpc/src/types/tx.rs

```diff
@@ -0,0 +1,245 @@
+    /// emitted from the ante. Whereas Logs contains the events, with
+    /// additional metadata, emitted only by processing the messages.
+    pub events: Vec<Event>,
+}
+
+/// Response to GetTx
+#[derive(Debug)]
+pub struct GetTxResponse {
+    /// Response Transaction
+    pub tx: Tx,
+
+    /// TxResponse to a Query
+    pub tx_response: TxResponse,
+}
+
+impl TryFrom<RawTxResponse> for TxResponse {
+    type Error = Error;
+
+    fn try_from(response: RawTxResponse) -> Result<TxResponse, Self::Error> {
+        Ok(TxResponse {
+            height: response.height,
+            txhash: response.txhash,
+            codespace: response.codespace,
+            code: response.code,
+            data: response.data,
+            raw_log: response.raw_log,
+            logs: response.logs,
+            info: response.info,
+            gas_wanted: response.gas_wanted,
+            gas_used: response.gas_used,
+            tx: response.tx,
+            timestamp: response.timestamp,
+            events: response.events,
+        })
+    }
+}
+
+impl FromGrpcResponse<TxResponse> for BroadcastTxResponse {
+    fn try_from_response(self) -> Result<TxResponse, Error> {
+        self.tx_response
+            .ok_or(Error::FailedToParseResponse)?
+            .try_into()
+    }
+}
+
+impl FromGrpcResponse<GetTxResponse> for RawGetTxResponse {
+    fn try_from_response(self) -> Result<GetTxResponse, Error> {
+        let tx_response = self
+            .tx_response
+            .ok_or(Error::FailedToParseResponse)?
+            .try_into()?;
+
+        let tx = self.tx.ok_or(Error::FailedToParseResponse)?;
+
+        let cosmrs_tx_body: cosmos_sdk_proto::cosmos::tx::v1beta1::TxBody =
+            tx.body.ok_or(Error::FailedToParseResponse)?.into();
+        let cosmrs_auth_info: cosmos_sdk_proto::cosmos::tx::v1beta1::AuthInfo =
+            tx.auth_info.ok_or(Error::FailedToParseResponse)?.into();
+
+        let cosmos_tx = Tx {
+            body: cosmrs_tx_body.try_into()?,
+            auth_info: cosmrs_auth_info.try_into()?,
+            signatures: tx.signatures,
+        };
+
+        Ok(GetTxResponse {
+            tx: cosmos_tx,
+            tx_response,
+        })
+    }
+}
+
+impl IntoGrpcParam<BroadcastTxRequest> for (RawTx, Vec<Blob>, BroadcastMode) {
+    fn into_parameter(self) -> BroadcastTxRequest {
+        let (tx, blobs, mode) = self;
+        assert!(blobs.len() > 0);
+        let blob_tx = new_blob_tx(&tx, blobs);
+        BroadcastTxRequest {
+            tx_bytes: blob_tx.encode_to_vec(),
+            mode: mode.into(),
+        }
+    }
+}
+
+impl IntoGrpcParam<RawGetTxRequest> for String {
+    fn into_parameter(self) -> RawGetTxRequest {
+        RawGetTxRequest { hash: self }
+    }
+}
+
+pub fn prep_signed_tx(
+    msg_pay_for_blobs: &MsgPayForBlobs,
+    base_account: &BaseAccount,
+    gas_limit: u64,
+    fee: u64,
+    chain_id: String,
+    account_keys: AccountKeypair,
```

Comment:
could we instead accept something like this?
```
public_key: VerifyingKey,
signer: impl Signer,
```

The reason for that is this enforces us to store sk there, while we can have different signers like hsm, or even use [api exposed by leap](https://docs.leapwallet.io/cosmos/for-dapps-connect-to-leap/api-reference#sign-direct-protobuf) for signing

Reply:
Got rid of the `AccountKeypair` struct (outside of tests) and switched to accepting public key and `impl Singer`